filebeat/input/filestream/input.go

```diff
@@ -177,12 +199,36 @@ func initState(log *logp.Logger, c loginp.Cursor, s fileSource) state {
 	return state
 }
 
-func (inp *filestream) open(log *logp.Logger, canceler input.Canceler, path string, offset int64) (reader.Reader, error) {
+func (inp *filestream) open(log *logp.Logger, canceler input.Canceler, srcID, path string, offset int64) (reader.Reader, error) {
 	f, err := inp.openFile(log, path, offset)
 	if err != nil {
 		return nil, err
 	}
 
+	if inp.detailedMetricsEnabled {
+		// if there was a previously running monitoring routine, stop it
+		if m, ok := inp.detailedMetrics[srcID]; ok {
+			m.cancel()
+		}
+
+		monitorCtx, monitorCancel := ctxtool.WithFunc(canceler, func() {
+			log.Debug("Stop monitoring of opened file")
+		})
+
+		inp.detailedMetrics[srcID] = newProgressMetrics(inp.ID, path, srcID, monitorCancel)
+		go func() {
+			timed.Periodic(monitorCtx, 30*time.Second, func() error {
+				fi, err := f.Stat()
```

Comment:
This go-routine is not cleaned up when the harvester is shutdown. That is, we will likely fail here because the file was already closed. Unfortunately we shouldn't keep the file open, in order to not exhaust file descriptors. Alternatively we might want the prospector/file watcher to update the size.

This reminds me. If possible, the reader should also update the internal 'size' one it hits EOF. That is, the reader should not read until EOF, but until the internal size. Once the limit is reached, using Stat to update the size and continue reading if the size is increased. This is e.g. helpful when reading from NFS, as the `stat` circumvents 0 reads, as it can force extra roundtrips at the protocol level.

Reply:
Why isn't it cleaned up? The canceler is from the `input.Context`, so if the parent is cancelled, it should be cancelled as well. What did I miss?

Reply:
Oh, I missed that the `monitorCtx` is derived from the input context. Nevermind.

There is still a potential race on `f`, as the overall input shutdown does not wait for the monitor to finish before the file is closed.